README.md
# Triangle mosaic

Generate colorful triangle-based SVG patterns (like this 👇) with ease.

<img src="https://raw.githubusercontent.com/bence-toth/triangle-mosaic/master/readme-assets/header.jpg" alt="Example output with triangles" style="width: 100%" />

## Demo

You can try Triangle mosaic [on the demo page](http://bence-toth.github.io/triangle-mosaic).


## Installation

To install `triangle-mosaic`, run:

```sh
npm install triangle-mosaic
```


## Quick start

You can import `TriangleMosaic` like this:

```js
import TriangleMosaic from 'triangle-mosaic'
```

`TriangleMosaic` is a class and it can be instantiated like this:

```js
const myMosaic = new TriangleMosaic(options)
```

Then you can call the `render` method which returns the SVG code as a string:

```js
const mySvgCode = myMosaic.render()
```

You can also change the options without losing the random-generated variance by calling the `rehydrate` method with the new options. This will also return the new SVG code as a string:

```js
const myNewSvgCode = myMosaic.rehydrate(newOptions)
```


## Options

The `options` parameter of both the constructor and the `rehydrate` method is an object with various options that influence the outcome.

If you haven’t yet, maybe it’s a good time to check out [the demo page](http://bence-toth.github.io/triangle-mosaic) to get a good overview of the features.


### Dimensions and number of tiles

| Member name    | Description                        |
|----------------|------------------------------------|
| `width`        | The width of the output in pixels  |
| `height`       | The height of the output in pixels |
| `xResolution`  | The number of tiles horizontally   |
| `yResolution`  | The number of tiles vertically     |

For example:

```js
const options = {
  ...,
  width:  640,
  height: 480,
  xResolution: 16,
  yResolution: 12,
  ...
}
```

The options `xResolution` and `yResolution` are the only options which cannot be rehydrated without losing the random-generated variance of tiles.


### Variance

| Member name            | Description                                                 |
|------------------------|-------------------------------------------------------------|
| `shapeFuzz`            | Variance factor of grid points (0-1)                        |
| `colorFuzz.hue`        | Variance factor of triangle color hue (0-1)                 |
| `colorFuzz.saturation` | Variance factor of triangle color saturation (0-1)          |
| `colorFuzz.lightness`  | Variance factor of triangle color lightness (0-1)           |
| `colorFuzz.alpha`      | Variance factor of triangle color alpha (0-1)               |
| `diagonals`            | Either `'nw-se'`, `'ne-sw'`, `'alternating'`, or `'random'` |

Variance options control how the triangles can deviate from their original shape and coloring.

The option `shapeFuzz` controls how far the grid points may venture from their original location.

Parameters that belong the object `colorFuzz` control how much the triangle colors may deviate from their original color. This can be controlled along all four axes of the HSLA color representation.

The option `diagonals` controls which diagonals of the squares of the grid are used:
- `'nw-se'` draws the diagonals from the top left corners to the bottom right corners (North-West - South-East)
- `'ne-sw'` draws the diagonals from the top right corners to the bottom left corners (North-East - South-West)
- `'alternating'` draws diagonals alternating, so that no two neighboring squares have the same diagonal
- `'random'` draws diagonals randomly

For example:

```js
const options = {
  ...,
  shapeFuzz: 0.65,
  colorFuzz: {
    hue:        0.1,
    saturation: 0.1,
    lightness:  0.1,
    alpha:      0
  },
  diagonals: 'nw-se',
  ...
}
```


### Coloring

The option `coloring` will decide the original color of the triangles. There are various possibilities here.


#### Coloring mode

| Member name     | Description   |
|-----------------|---------------|
| `coloring.mode` | Coloring mode |

The option `coloring.mode` will decide how the rest of the `coloring` object is interpreted.

The following coloring modes are supported:
- Single color (`single`)
- Linear gradient (`linearGradient`)
- Radial gradient (`radialGradient`)
- Color spots (`spots`)

Their additional options of the `coloring` object are detailed below for each of the coloring modes.


#### Single color

| Member name      | Description                     |
|------------------|---------------------------------|
| `coloring.color` | Base color (hexadecimal format) |

The single color mode only takes a color in hexadecimal color format:

```js
const options = {
  ...,
  coloring: {
    mode:  'single',
    color: '#ffc107'
  },
  ...
}
```


#### Linear and radial gradient

| Member name      | Description                                                                       |
|------------------|-----------------------------------------------------------------------------------|
| `coloring.start` | An object with the location of the gradient’s starting point                      |
| `coloring.end`   | An object with the location of the gradient’s end point                           |
| `coloring.stops` | An array describing all stops of the gradient.                                    |

The linear and radial gradient color modes require three additional options in the `coloring` object:

The options `coloring.start` and `coloring.end` are object that have members `x` and `y` containing the horizontal and vertical coordinates of the gradient’s starting and end points.

The `coloring.stops` option is an array of arrays, where the elements have two items: a location (0-1) and a color (in hexadecimal color format). The locations of the first and the last spot must always be 0 and 1, respectively.

For example:

```js
const options = {
  ...,
  coloring: {
    mode: 'linearGradient', // or 'radialGradient'
    start: {
      x: 0,
      y: 0
    },
    end: {
      x: 1280,
      y: 720
    },
    stops: [
      [0,    '#9c27b0'],
      [0.25, '#03a9f4'],
      [0.5,  '#8bc34a'],
      [0.75, '#ffc107'],
      [1,    '#f44336']
    ]
  },
  ...
}
```


#### Spots

| Member name              | Description                                    |
|--------------------------|------------------------------------------------|
| `coloring.spots`         | An array of objects describing all color spots |
| `coloring.spotIntensity` | The default intensity of spots                 |


The `coloring.spots` option is an array of color spots, where the elements have four members:

- `x` and `y` containing the horizontal and vertical coordinates of the color spot

- `color` containing the spot’s color in hexadecimal color format

- `intensity` which represents how much this light spot contributes to individual triangle colors as compared to other spots (0-1)

The `coloring.spotIntensity` (0-1) option sets the intensity for spots that don’t have their individual intensity specified.

For example:

```js
const options = {
  ...,
  coloring: {
    mode: 'spots',
    spotIntensity: 0.5,
    spots: [
      {
        x: 0,
        y: 0,
        color: '#ffc107',
        intensity: 0.65
      },
      {
        x: 1280,
        y: 0,
        color: '#f44336'
        // intensity falls back to 0.5
      },
      {
        x: 640,
        y: 720,
        color: '#2196f3',
        intensity: 0.6
      }
    ]
  },
  ...
}
```


### Full example

```js
import TriangleMosaic from 'triangle-mosaic'

let options = {
  width:  640,
  height: 480,
  xResolution: 12,
  yResolution: 8,
  shapeFuzz: 0.65,
  colorFuzz: {
    hue:        0.1,
    saturation: 0.1,
    lightness:  0.1,
    alpha:      0
  },
  diagonals: 'nw-se',
  coloring: {
    mode: 'linearGradient',
    start: {
      x: 0,
      y: 0
    },
    end: {
      x: 640,
      y: 480
    },
    stops: [
      {
        id: 0,
        location: 0,
        color: '#9c27b0'
      },
      {
        id: 1,
        location: 0.25,
        color: '#03a9f4'
      },
      {
        id: 2,
        location: 0.5,
        color: '#8bc34a'
      },
      {
        id: 3,
        location: 0.75,
        color: '#ffc107'
      },
      {
        id: 4,
        location: 1,
        color: '#f44336'
      }
    ]
  }
}

// Render SVG
const myMosaic = new TriangleMosaic(options)
const mySvgCode = myMosaic.render()

console.log(mySvgCode)

// Change options and render SVG again
options.width = 720
options.coloring.end.x = 720
options.shapeFuzz = 0.5
options.colorFuzz.hue = 0

const myNewSvgCode = myMosaic.rehydrate(options)

console.log(myNewSvgCode)
```


## License

Triangle Mosaic is [licensed under MIT](./LICENSE).

The images you create with the library or web-based tool are yours and you may license them in any way you see fit.


changelog.md
# Change log

## 1.1.0

Bugfix: Gradients do not fail anymore if start and end points have the same x coordinates.

Bugfix: Hue fuzz/variance can now correctly cross the 0-360° mark.

New feature: Introduced `diagonal` option.


## 1.0.1

Got rid of optional chaining operator as it was causing issues.


## 1.0.0

Initial release.


triangle-mosaic/triangle-mosaic.js
// ----------------------------------------------------------------------------
// Render

const renderSvg = ({
  width,
  height,
  children
} = {}) => `
  <svg width="${width}" height="${height}">
    ${children}
  </svg>
`

const renderTriangle = ({
  edges,
  color
}) => `
  <polygon
    points="
      ${edges[0].x},${edges[0].y}
      ${edges[1].x},${edges[1].y}
      ${edges[2].x},${edges[2].y}"
    style="
      fill: ${color};
      stroke: ${color};
      stroke-width: 1;
    "
  />
`

// ----------------------------------------------------------------------------
// Utility

const movePoint = ({
  x,
  y,
  direction,
  distance,
  ...rest
}) => ({
  x: x + (distance * Math.cos(direction)),
  y: y + (distance * Math.sin(direction)),
  ...rest
})

const getDistance = (
  {x: x1, y: y1},
  {x: x2, y: y2}
) => (
  Math.sqrt(
    ((x2 - x1) ** 2) +
    ((y2 - y1) ** 2)
  )
)

const hexToRgb = hex => {
  const result = /^#?([\da-f]{2})([\da-f]{2})([\da-f]{2})$/i.exec(hex)
  return result ? {
    r: parseInt(result[1], 16),
    g: parseInt(result[2], 16),
    b: parseInt(result[3], 16)
  } : null
}

const getRgbaColor = ({r, g, b, a}) => (
  (a === 1)
    ? `rgb(${r}, ${g}, ${b})`
    : `rgba(${r}, ${g}, ${b}, ${a})`
)

const add = (a, b) => a + b

const getAverage = (...numbers) => (
  numbers.reduce(add) / numbers.length
)

const getTriangleCenter = (
  [
    {x: x1, y: y1},
    {x: x2, y: y2},
    {x: x3, y: y3}
  ]
) => ({
  x: getAverage(x1, x2, x3),
  y: getAverage(y1, y2, y3)
})

const modulo = (x, n) => (x % n + n) % n;

const rotate = ({min, value, max}) => (
  min + (modulo(value - min, max - min))
)

const clamp = ({min, value, max}) => (
  Math.max(min, Math.min(max, value))
)

const getPerpendicularPoint = ({
  start: s,
  end: e,
  external: t
}) => {
  const a = t.x - s.x
  const b = t.y - s.y
  const c = e.x - s.x
  const d = e.y - s.y
  const dot = (a * c) + (b * d)
  const lengthSquared = (c ** 2) + (d ** 2)
  const projectionLength = (
    (lengthSquared !== 0)
      ? (dot / lengthSquared)
      : -1
  )
  return {
    x: s.x + (projectionLength * c),
    y: s.y + (projectionLength * d)
  }
}

const getRatio = (start, end, target) => (
  (target - start) / (end - start)
)

const hueToRgb = (p, q, t) => {
  if (t < 0) {
    t += 1
  }
  if (t > 1) {
    t -= 1
  }
  if (t < 1/6) {
    return p + (q - p) * 6 * t
  }
  if (t < 1/2) {
    return q
  }
  if (t < 2/3) {
    return p + (q - p) * (2/3 - t) * 6
  }
  return p
}

const hslaToRgba = ({h, s, l, a}) => {
  let r, g, b

  if (s === 0) {
    r = g = b = l
  }
  else {
    const q = (l < 0.5) ? (l * (1 + s)) : (l + s - (l * s))
    const p = (2 * l) - q
    r = hueToRgb(p, q, h + 1/3)
    g = hueToRgb(p, q, h)
    b = hueToRgb(p, q, h - 1/3)
  }

  return {
    r: r * 255,
    g: g * 255,
    b: b * 255,
    a
  }
}

const rgbToHsl = ({r, g, b}) => {
  r /= 255
  g /= 255
  b /= 255
  const max = Math.max(r, g, b)
  const min = Math.min(r, g, b)
  let h, s, l = (max + min) / 2

  if (max === min) {
    h = s = 0
  }
  else {
    const d = max - min
    s = (l > 0.5) ? (d / (2 - max - min)) : (d / (max + min))

    switch (max) {
      case r:
        h = ((g - b) / d) + ((g < b) ? 6 : 0)
        break
      case g:
        h = ((b - r) / d) + 2
        break
      case b:
        h = ((r - g) / d) + 4
        break
    }

    h /= 6
  }

  return {h, s, l}
}

const adjustColor = ({color, adjustments}) => {
  const {h, s, l} = rgbToHsl(color)
  const adjustedHslaColor = {
    h: rotate({
      min: 0,
      max: 1,
      value: h + (adjustments.hue * 0.15)
    }),
    s: clamp({
      min: 0,
      max: 1,
      value: s * (adjustments.saturation + 1)
    }),
    l: clamp({
      min: 0,
      max: 1,
      value: l * (adjustments.lightness + 1)
    }),
    a: clamp({
      min: 0,
      max: 1,
      value: 1 - adjustments.alpha
    })
  }
  return hslaToRgba(adjustedHslaColor)
}

const getTriangleColorForGradient = ({coloring, center}) => {
  const {start, end, stops, mode} = coloring
  let ratio

  if (mode === 'linearGradient') {
    const perpendicularPoint = getPerpendicularPoint({
      start,
      end,
      external: center
    })

    ratio = (
      (start.x !== end.x)
        ? getRatio(start.x, end.x, perpendicularPoint.x)
        : getRatio(start.y, end.y, perpendicularPoint.y)
    ) || 0
  }

  if (mode === 'radialGradient') {
    const gradientLength = getDistance(start, end)
    const triangleDistance = getDistance(start, center)

    ratio = (triangleDistance / gradientLength) || 0
  }

  const exactMatch = stops.find(([location]) => ratio === location)

  if (ratio < 0) {
    // Went negative overboard
    return hexToRgb(stops[0][1])
  }
  if (ratio > 1) {
    // Went positive overboard
    return hexToRgb(stops[stops.length - 1][1])
  }
  if (exactMatch) {
    // Hit a stop exactly
    return hexToRgb(exactMatch[1])
  }
  // Somewhere between two stops
  const nextStopIndex = stops.findIndex(([location]) => location > ratio)
  const previousStop = stops[nextStopIndex - 1]
  const nextStop = stops[nextStopIndex]
  const endRatio = getRatio(previousStop[0], nextStop[0], ratio)
  const endColor = hexToRgb(nextStop[1])
  const startRatio = 1 - endRatio
  const startColor = hexToRgb(previousStop[1])

  return {
    r: (startRatio * startColor.r) + (endRatio * endColor.r),
    g: (startRatio * startColor.g) + (endRatio * endColor.g),
    b: (startRatio * startColor.b) + (endRatio * endColor.b)
  }
}

const getTriangleColorForSpots = ({coloring, center}) => {
  const {spots, spotIntensity} = coloring

  const getWeight = spot => (
    1 / (getDistance(spot, center) ** (1 / (spot.intensity || spotIntensity || 0.5)))
  )

  const fullWeight = (
    spots
      .map(getWeight)
      .reduce(add, 0)
  )

  return (
    spots
      .map(spot => ({
        // Get color components of the spot
        color: hexToRgb(spot.color),
        // Calculate how much this spot contributes to the color
        factor: getWeight(spot) / fullWeight
      }))
      .map(({
        color: {r, g, b},
        factor
      }) => ({
        // Calculate the color this spot contributes to the mix with
        r: r * factor,
        g: g * factor,
        b: b * factor
      }))
      .reduce((accumulator, currentValue) => ({
        // Add color values of all spots (by color component)
        r: accumulator.r + currentValue.r,
        g: accumulator.g + currentValue.g,
        b: accumulator.b + currentValue.b
      }), {r: 0, g: 0, b: 0})
  )
}

const getTriangleColor = ({
  triangle,
  colorFuzz,
  colorDeviation,
  coloring
}) => {
  let color = {r: 0, g: 0, b: 0}

  if (coloring.mode === 'single') {
    color = hexToRgb(coloring.color)
  }

  const center = getTriangleCenter(triangle)

  if (['linearGradient', 'radialGradient'].includes(coloring.mode)) {
    color = getTriangleColorForGradient({coloring, center})
  }
  if (coloring.mode === 'spots') {
    color = getTriangleColorForSpots({coloring, center})
  }

  const adjustedColor = adjustColor({
    color,
    adjustments: {
      hue: colorDeviation.hue * colorFuzz.hue,
      saturation: colorDeviation.saturation * colorFuzz.saturation,
      lightness: colorDeviation.lightness * colorFuzz.lightness,
      alpha: colorDeviation.alpha * colorFuzz.alpha
    }
  })

  return adjustedColor
}

const getMaxVentureDistance = ({
  width,
  height,
  xResolution,
  yResolution
}) => {
  const horizontalDistance = height / yResolution
  const verticalDistance = width / xResolution
  const smallerDistance = Math.min(horizontalDistance, verticalDistance)
  const maxVentureDistance = smallerDistance / 2
  return maxVentureDistance
}

const getRandomTriangleColorDeviation = () => ({
  hue: (Math.random() * 2) - 1,
  saturation: (Math.random() * 2) - 1,
  lightness: (Math.random() * 2) - 1,
  alpha: Math.random()
})

const getRandomGridPointOptions = () => ({
  direction: Math.random() * Math.PI * 2,
  factor: Math.random(),
  diagonal: Math.round(Math.random()),
  topTriangleColorDeviation: getRandomTriangleColorDeviation(),
  bottomTriangleColorDeviation: getRandomTriangleColorDeviation()
})

// ----------------------------------------------------------------------------
// Engine

const getGrid = ({
  xResolution,
  yResolution
}) => {
  const numberOfRows = yResolution + 1
  const numberOfColumns = xResolution + 1

  const gridPoints = []

  for (let rowCounter = 0; rowCounter < numberOfRows; ++rowCounter) {
    const gridPointsInRow = []
    for (let columnCounter = 0; columnCounter < numberOfColumns; ++columnCounter) {
      gridPointsInRow.push({
        x: columnCounter,
        y: rowCounter,
        ...getRandomGridPointOptions()
      })
    }
    gridPoints.push(gridPointsInRow)
  }

  return gridPoints
}

const getTriangles = ({
  grid,
  shapeFuzz,
  colorFuzz,
  coloring,
  diagonals,
  width,
  height
}) => {
  const numberOfRows = grid.length
  const numberOfColumns = grid[0].length
  const verticalDistance = height / (numberOfRows - 1)
  const horizontalDistance = width / (numberOfColumns - 1)

  const triangles = []

  // We'll take 4 grid points in one go to form 2 triangles
  for (let rowCounter = 0; rowCounter < numberOfRows - 1; ++rowCounter) {
    for (let columnCounter = 0; columnCounter < numberOfColumns - 1; ++columnCounter) {
      const isFirstRow = (rowCounter === 0)
      const isFirstColumn = (columnCounter === 0)
      const isLastRow = (rowCounter === numberOfRows - 2)
      const isLastColumn = (columnCounter === numberOfColumns - 2)

      const isPointStatic = pointPosition => {
        if (pointPosition === 0) {
          return isFirstRow || isFirstColumn
        }
        if (pointPosition === 1) {
          return isFirstRow || isLastColumn
        }
        if (pointPosition === 2) {
          return isFirstColumn || isLastRow
        }
        if (pointPosition === 3) {
          return isLastRow || isLastColumn
        }
      }

      const points = [
        //
        // 0----1    0----1
        // |  / | or | \  |
        // | /  |    |  \ |
        // 2----3    2----3
        //
        grid[rowCounter][columnCounter],
        grid[rowCounter][columnCounter + 1],
        grid[rowCounter + 1][columnCounter],
        grid[rowCounter + 1][columnCounter + 1]
      ].map((point, pointIndex) => ({
        ...point,
        x: point.x * horizontalDistance,
        y: point.y * verticalDistance,
        direction: point.direction,
        distance: (
          !isPointStatic(pointIndex)
            ? (point.factor * shapeFuzz)
            : 0
        )
      })).map(movePoint)

      const {
        topTriangleColorDeviation,
        bottomTriangleColorDeviation,
        diagonal: pointDiagonal
      } = points[0]

      const diagonal = (() => {
        if (diagonals === 'nw-se') {
          return 0
        }
        if (diagonals === 'ne-sw') {
          return 1
        }
        if (diagonals === 'alternating') {
          return (
            (
              ((rowCounter % 2) === 0) && ((columnCounter % 2) === 1)
                || ((rowCounter % 2) === 1) && ((columnCounter % 2) === 0)
            )
              ? 0
              : 1
          )
        }
        if (diagonals === 'random') {
          return pointDiagonal
        }
        return 0
      })()

      const topTriangle = {
        // Top triangle
        //
        // 0----1    0----1
        // |  /   or   \  |
        // | /          \ |
        // 2              3
        //
        edges: (
          (diagonal === 0)
            ? [points[0], points[1], points[2]]
            : [points[0], points[1], points[3]]
        ),
        color: getRgbaColor(getTriangleColor({
          triangle: (
            (diagonal === 0)
              ? [points[0], points[1], points[2]]
              : [points[0], points[1], points[3]]
          ),
          coloring,
          colorFuzz,
          colorDeviation: topTriangleColorDeviation
        }))
      }

      const bottomTriangle = {
        // Bottom triangle
        //
        //      1    0
        //    / | or | \
        //   /  |    |  \
        // 2----3    2----3
        //
        edges: (
          (diagonal === 0)
            ? [points[1], points[3], points[2]]
            : [points[0], points[3], points[2]]
        ),
        color: getRgbaColor(getTriangleColor({
          triangle: (
            (diagonal === 0)
              ? [points[1], points[3], points[2]]
              : [points[0], points[3], points[2]]
          ),
          coloring,
          colorFuzz,
          colorDeviation: bottomTriangleColorDeviation
        }))
      }

      triangles.push(topTriangle, bottomTriangle)
    }
  }

  return triangles
}

// ----------------------------------------------------------------------------
// State and API

class TriangleMosaic {
  constructor({
    width = 1280,
    height = 720,
    xResolution = 16,
    yResolution = 9,
    shapeFuzz = 0.65,
    colorFuzz = {
      hue: 0.1,
      saturation: 0.1,
      lightness: 0.1,
      alpha: 0
    },
    diagonals = 'ne-sw',
    coloring = {
      mode: 'spots',
      spots: [
        {
          x: 0,
          y: 0,
          color: '#ffc107'
        },
        {
          x: 1280,
          y: 0,
          color: '#f44336'
        },
        {
          x: 640,
          y: 720,
          color: '#2196f3'
        }
      ]
    }
  } = {}) {
    this.height = height
    this.width = width
    this.xResolution = xResolution
    this.yResolution = yResolution
    this.shapeFuzz = shapeFuzz
    this.colorFuzz = colorFuzz
    this.diagonals = diagonals
    this.coloring = coloring
    this.grid = getGrid({
      width,
      height,
      xResolution,
      yResolution
    })
  }

  render() {
    const maxVentureDistance = getMaxVentureDistance({
      width: this.width,
      height: this.height,
      xResolution: this.xResolution,
      yResolution: this.yResolution
    })
    const triangles = getTriangles({
      grid: this.grid,
      shapeFuzz: this.shapeFuzz * maxVentureDistance,
      colorFuzz: this.colorFuzz,
      diagonals: this.diagonals,
      coloring: this.coloring,
      width: this.width,
      height: this.height
    })
    return renderSvg({
      width: this.width,
      height: this.height,
      children: triangles.map(renderTriangle).join('')
    })
  }

  rehydrate({
    shapeFuzz,
    colorFuzz,
    coloring,
    width,
    height,
    diagonals,
    xResolution,
    yResolution,
  } = {}) {
    if (shapeFuzz !== undefined) {
      this.shapeFuzz = shapeFuzz
    }
    if (colorFuzz && (colorFuzz.hue !== undefined)) {
      this.colorFuzz.hue = colorFuzz.hue
    }
    if (colorFuzz && (colorFuzz.saturation !== undefined)) {
      this.colorFuzz.saturation = colorFuzz.saturation
    }
    if (colorFuzz && (colorFuzz.lightness !== undefined)) {
      this.colorFuzz.lightness = colorFuzz.lightness
    }
    if (colorFuzz && (colorFuzz.alpha !== undefined)) {
      this.colorFuzz.alpha = colorFuzz.alpha
    }
    if (coloring !== undefined) {
      this.coloring = coloring
    }
    if (width !== undefined) {
      this.width = width
    }
    if (height !== undefined) {
      this.height = height
    }
    if (diagonals !== undefined) {
      this.diagonals = diagonals
    }
    if (xResolution !== undefined) {
      this.xResolution = xResolution
    }
    if (yResolution !== undefined) {
      this.yResolution = yResolution
    }
    if (xResolution !== undefined || yResolution !== undefined) {
      this.grid = getGrid({
        width: this.width,
        height: this.height,
        xResolution: this.xResolution,
        yResolution: this.yResolution
      })
    }
    return this.render()
  }
}

export default TriangleMosaic
